martian/src/lib.rs

```diff
@@ -40,74 +36,33 @@ pub mod mro;
 pub use mro::*;
 
 pub use log::LevelFilter;
-
 pub mod prelude;
 
-/// Ways a stage can fail.
-///
-/// The failure can either be a controlled failure or an unexpected
-/// failure.
-#[derive(Debug, Fail)]
-pub enum StageError {
-    /// Controlled shutdown for known condition in data or config
-    #[fail(display = "{}", message)]
-    MartianExit { message: String },
-
-    /// Unexpected error
-    #[fail(display = "{}", message)]
-    PipelineError { message: String },
-}
 
-pub fn initialize(args: Vec<String>, log_file: &File) -> Result<Metadata, Error> {
+pub fn initialize(args: Vec<String>, log_file: File) -> Result<Metadata, Error> {
     let mut md = Metadata::new(args, log_file);
     println!("got metadata: {:?}", md);
     md.update_jobinfo()?;
 
     Ok(md)
 }
 
-pub fn handle_stage_error(err: Error) {
-    // Try to handle know StageError cases
-    match err.downcast::<StageError>() {
-        Ok(ref e) => {
-            match *e {
-                StageError::MartianExit { message: ref m } => {
-                    let _ = write_errors(&format!("ASSERT: {}", m));
-                }
-                // No difference here at this point
-                StageError::PipelineError { message: ref m } => {
-                    let _ = write_errors(&format!("ASSERT: {}", m));
-                }
-            }
-        }
-        Err(ref e) => {
-            let msg = format!("stage error:{}\n{}", e.as_fail(), e.backtrace());
-            let _ = write_errors(&msg);
-        }
-    }
-}
-
-fn write_errors(msg: &str) -> Result<(), Error> {
-    unsafe {
-        let mut err_file = File::from_raw_fd(4);
-        let _ = err_file.write(msg.as_bytes())?;
-        Ok(())
-    }
-}
-
-/// Log a panic to the martian output machinery
-pub fn log_panic(panic: &panic::PanicInfo) {
-    let payload = match panic.payload().downcast_ref::<String>() {
-        Some(as_string) => as_string.to_string(),
-        None => format!("{:?}", panic.payload()),
+fn write_errors(msg: &str, is_assert: bool) -> Result<(), Error> {
```

Comment:
Does err_file going out of scope cause the file to be closed?  If so, then this method should be `-> !` and should end by flushing the file and then exiting the process.  `mrjob` will `kill -9` you once the file is closed anyway, so it's dangerous to count on being able to do anything afterwards anyway.  There's also a bunch of cool optimizations the compiler can do on such functions, if it knows the stack is never going to get popped.
```suggestion
fn write_errors(msg: &str, is_assert: bool) -> ! {
```

Reply:
It scares me a bit for the process to be killed, so I've changed this to keep the file handle open.